### ISSUE TITLE
Lazy tests failing due to KeyError in dask.optimize

### ISSUE THREAD

**Issue by jan-car** [hyperspy/hyperspy#1525]
Taken from #1522:

When testing HyperSpy, I get around five errors, all in lazy tests, e.g. in `test_integrate_in_range.py` in line 23:
```python
def test_integrate_in_range(self):
        integrated_signal = self.signal.integrate_in_range(signal_range=(None, None))
>       assert np.allclose(integrated_signal.data, 20,)
signal\test_integrate_in_range.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\Jan\Anaconda3\lib\site-packages\numpy\core\numeric.py:2478: in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
C:\Users\Jan\Anaconda3\lib\site-packages\numpy\core\numeric.py:2548: in isclose
    x = array(a, copy=False, subok=True, ndmin=1)
C:\Users\Jan\Anaconda3\lib\site-packages\dask\array\core.py:898: in __array__
    x = self.compute()
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:86: in compute
    return compute(self, **kwargs)[0]
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:175: in compute
    for opt, val in groups.items()])
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:175: in <listcomp>
    for opt, val in groups.items()])
C:\Users\Jan\Anaconda3\lib\site-packages\dask\array\optimization.py:22: in optimize
    dsk2, dependencies = cull(dsk, keys)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dsk = {('array-946e3825e554e085d547a7cd0b0d03ee', 0): (<function getarray at 0x00000000058D3C80>, 'array-original-946e3825e5...2.97343903e-06,   2.98834196e-06,   3.00331658e-06, ...,
         3.01836322e-06,   3.00331658e-06,   2.98834196e-06])}
keys = [('simps-8871a51f818b0c69babffd65b8879f58', 0)]

    def cull(dsk, keys):
        """ Return new dask with only the tasks required to calculate keys.
    
        In other words, remove unnecessary tasks from dask.
        ``keys`` may be a single key or list of keys.
    
        Examples
        --------
        >>> d = {'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)}
        >>> dsk, dependencies = cull(d, 'out')  # doctest: +SKIP
        >>> dsk  # doctest: +SKIP
        {'x': 1, 'out': (add, 'x', 10)}
        >>> dependencies  # doctest: +SKIP
        {'x': set(), 'out': set(['x'])}
    
        Returns
        -------
        dsk: culled dask graph
        dependencies: Dict mapping {key: [deps]}.  Useful side effect to accelerate
            other optimizations, notably fuse.
        """
        if not isinstance(keys, (list, set)):
            keys = [keys]
        out = dict()
        seen = set()
        dependencies = dict()
        stack = list(set(flatten(keys)))
        while stack:
            key = stack.pop()
>           out[key] = dsk[key]
E           KeyError: ('simps-8871a51f818b0c69babffd65b8879f58', 0)

C:\Users\Jan\Anaconda3\lib\site-packages\dask\optimize.py:43: KeyError
```

Not sure what the problem is, but @magnunor apparently has similar issues.

**Comment by jan-car** [comment]
A small update:
Upgrading dask from `0.11.0-py35_0 --> 0.14.0-py35_0` did not help...

I don't get the errors mentioned above, but instead  errors in `test_lazy.py` for `test_reshuffle` and `test_blockiter_bothmasks`, which look like this:

```python
signal = <LazySignal2D, title: , dimensions: (9, 6|11, 7)>, flat = True
dtype = 'float', nm = None, sm = None

    @pytest.mark.parametrize('nm', [None, nav_mask])
    @pytest.mark.parametrize('sm', [None, sig_mask])
    @pytest.mark.parametrize('flat', [True, False])
    @pytest.mark.parametrize('dtype', ['float', 'int'])
    def test_blockiter_bothmasks(signal, flat, dtype, nm, sm):
>       real_first = get(signal.data.dask, (signal.data.name, 0, 0, 0, 0)).copy()

signal\test_lazy.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\Jan\Anaconda3\lib\site-packages\dask\threaded.py:76: in get
    **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

apply_async = <bound method Pool.apply_async of <multiprocessing.pool.ThreadPool object at 0x000000000D3F3AC8>>
num_workers = 4, dsk = <dask.sharedict.ShareDict object at 0x00000000191E6F98>
result = ('array-7b7fd7cb81c41c99c8668108e4090224', 0, 0, 0, 0), cache = None
get_id = <function _thread_get_id at 0x00000000061986A8>
raise_on_exception = False, rerun_exceptions_locally = None, callbacks = set()
dumps = <function identity at 0x000000000612A0D0>
loads = <function identity at 0x000000000612A0D0>, kwargs = {}, _ = ()
posttask_cbs = ()
result_flat = {('array-7b7fd7cb81c41c99c8668108e4090224', 0, 0, 0, 0)}
results = {('array-7b7fd7cb81c41c99c8668108e4090224', 0, 0, 0, 0)}

    def get_async(apply_async, num_workers, dsk, result, cache=None,
                  get_id=default_get_id, raise_on_exception=False,
                  rerun_exceptions_locally=None, callbacks=None,
                  dumps=identity, loads=identity,
                  **kwargs):
        """ Asynchronous get function
    
        This is a general version of various asynchronous schedulers for dask.  It
        takes a an apply_async function as found on Pool objects to form a more
        specific ``get`` method that walks through the dask array with parallel
        workers, avoiding repeat computation and minimizing memory use.
    
        Parameters
        ----------
    
        apply_async : function
            Asynchronous apply function as found on Pool or ThreadPool
        num_workers : int
            The number of active tasks we should have at any one time
        dsk: dict
            A dask dictionary specifying a workflow
        result : key or list of keys
            Keys corresponding to desired data
        cache : dict-like, optional
            Temporary storage of results
        get_id : callable, optional
            Function to return the worker id, takes no arguments. Examples are
            `threading.current_thread` and `multiprocessing.current_process`.
        rerun_exceptions_locally : bool, optional
            Whether to rerun failing tasks in local process to enable debugging
            (False by default)
        dumps: callable, optional
            Function to serialize task data and results to communicate between
            worker and parent.  Defaults to identity.
        loads: callable, optional
            Inverse function of `dumps`.  Defaults to identity.
        callbacks : tuple or list of tuples, optional
            Callbacks are passed in as tuples of length 5. Multiple sets of
            callbacks may be passed in as a list of tuples. For more information,
            see the dask.diagnostics documentation.
    
        See Also
        --------
    
        threaded.get
        """
        queue = Queue()
    
        if callbacks is None:
            callbacks = _globals['callbacks']
        _, _, pretask_cbs, posttask_cbs, _ = unpack_callbacks(callbacks)
    
        if isinstance(result, list):
            result_flat = set(flatten(result))
        else:
            result_flat = set([result])
        results = set(result_flat)
    
>       dsk = dsk.copy()
E       AttributeError: 'ShareDict' object has no attribute 'copy'

C:\Users\Jan\Anaconda3\lib\site-packages\dask\async.py:444: AttributeError
```

**Comment by jan-car** [comment]
Updating to dask version `0.13.0` apparently solved this problem and also #1523, I'll close both!
